cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/RunnableTaskLauncher.java

```diff
@@ -16,36 +16,76 @@
 
 package io.cdap.cdap.internal.app.worker;
 
+import com.google.inject.Guice;
+import com.google.inject.Injector;
 import io.cdap.cdap.common.conf.CConfiguration;
-import io.cdap.cdap.common.internal.worker.RunnableTask;
-import io.cdap.cdap.common.internal.worker.RunnableTaskContext;
+import io.cdap.cdap.common.id.Id;
+import io.cdap.cdap.internal.app.runtime.artifact.ArtifactDetail;
+import io.cdap.cdap.internal.app.runtime.artifact.ArtifactRepository;
+import io.cdap.cdap.internal.app.runtime.artifact.ArtifactRepositoryReader;
+import io.cdap.cdap.internal.worker.api.RunnableTask;
+import io.cdap.cdap.internal.worker.api.RunnableTaskContext;
+import io.cdap.cdap.internal.worker.api.RunnableTaskRequest;
+import io.cdap.cdap.security.impersonation.EntityImpersonator;
+import io.cdap.cdap.security.impersonation.Impersonator;
+import org.apache.hadoop.conf.Configuration;
 
 /**
  * RunnableTaskLauncher launches a {@link RunnableTask} by loading its class and calling its run method.
  */
 public class RunnableTaskLauncher {
-  private final CConfiguration cConf;
+  private final CConfiguration cConfig;
+  private final Configuration hConf;
+  private final ArtifactRepositoryReader artifactRepositoryReader;
+  private final ArtifactRepository artifactRepository;
+  private final Impersonator impersonator;
 
-  public RunnableTaskLauncher(CConfiguration cConf) {
-    this.cConf = cConf;
+  public RunnableTaskLauncher(CConfiguration cConfig,
+                              Configuration hConf,
+                              ArtifactRepositoryReader artifactRepositoryReader,
+                              ArtifactRepository artifactRepository,
+                              Impersonator impersonator) {
+    this.cConfig = cConfig;
+    this.hConf = hConf;
+    this.artifactRepositoryReader = artifactRepositoryReader;
+    this.artifactRepository = artifactRepository;
+    this.impersonator = impersonator;
   }
 
   public byte[] launchRunnableTask(RunnableTaskRequest request) throws Exception {
-
-    ClassLoader classLoader = Thread.currentThread().getContextClassLoader();
-    if (classLoader == null) {
-      classLoader = getClass().getClassLoader();
-    }
+    ClassLoader classLoader = getClassLoader(request.getArtifactId(), request.getNamespace());
+    Injector injector = getInjector(request.getArtifactId() != null);
```

Comment:
I don't think we should have the logic here. The Task being prepared by the caller should create the Task that contains everything it needs.

Reply:
Since the task was part of the system app, this was required to load the task itself.

Reply:
The task I am referring to is not the task from the system app. The system app created a task, which should get wrapped by another task in app fabric, which do all these artifact and classloader logic before it submit to the task runner

Reply:
Makes sense. Moved this code to SystemAppTask .